### PR TITLE
fix expectations around kinesis record encoding

### PR DESCRIPTION
**Pull request by SwatsonCodes**
This PR should fix https://github.com/localstack/localstack/issues/4095. It took me a little while to figure out what was causing boto to throw a base64 decode error when retrieving records from Kinesis, but ultimately I identified two problems:

* The Kinesis tests make faulty assumptions about how kinesis/boto handle record encoding/decoding
* The `subscribe_to_shard` function has a special case for re-encoding records when talking to a Java client, but in fact this should happen always.

According to the [Kinesis docs](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecordsRequestEntry.html), Kinesis expects records to consist of base64 encoded data when you `PUT` them to the API. However, boto takes care of this for you behind the scenes: it automatically base64 encodes your data before it sends it to the API, and automatically decodes it when you retrieve the records. To sanity check this, I wrote this simple script for creating a stream, putting a record, and retrieving it.
```
import boto3
from time import sleep
boto3.set_stream_logger(name='botocore')

client = boto3.client('kinesis')
client.create_stream(StreamName='test', ShardCount=1)
sleep(10)
client.put_record(StreamName='test', Data='this is raw data', PartitionKey='foo')

iterator = client.get_shard_iterator( StreamName="test", ShardId='shardId-000000000000', ShardIteratorType='TRIM_HORIZON')["ShardIterator"]

result = client.get_records(ShardIterator=iterator)
print(result)
```
I made boto do verbose logging so that we can see the request and response bodies. I won't include the full output here because it includes sensitive data, but here are some snippets that illustrate that the record data is automatically encoded before it is sent and after it is retrieved:
```
[...]
# sending record, note that the data is automatically encoded
2022-01-13 12:33:21,902 botocore.hooks [DEBUG] Event before-call.kinesis.PutRecord: calling handler <function inject_api_version_header_if_needed at 0x11003cb80>
2022-01-13 12:33:21,902 botocore.endpoint [DEBUG] Making request for OperationModel(name=PutRecord) with params: {'url_path': '/', 'query_string': '', 'method': 'POST', 'headers': {...}, 'body': b'{"StreamName": "test", "Data": "dGhpcyBpcyByYXcgZGF0YQ==", "PartitionKey": "foo"}', 'url': 'https://kinesis.us-east-1.amazonaws.com/', 'context': {...}}

[...]

# API response for getting record
2022-01-13 12:33:23,011 botocore.parsers [DEBUG] Response body:
b'{"MillisBehindLatest":0,"NextShardIterator":"...","Records":[{"ApproximateArrivalTimestamp":1.642102402844E9,"Data":"dGhpcyBpcyByYXcgZGF0YQ==","PartitionKey":"foo","SequenceNumber":"..."}]}'

# finally printing the resulting records at the end of the script. Note how boto automatically decoded the record data.
{'Records': [{'SequenceNumber': '...', 'ApproximateArrivalTimestamp': '...', 'Data': b'this is raw data', 'PartitionKey': 'foo'}], 'NextShardIterator': ...}
```

Judging by the way the tests were written, it seems like we assumed that boto would automatically decode record data blobs, but not automatically encode them. I fixed this and added a more comprehensive test to ensure that unicode can be properly encoded/decoded as well.

Finally, it seems like this bug sort of surfaced in the past because of [this](https://github.com/localstack/localstack/blob/master/localstack/services/kinesis/kinesis_listener.py#L338) special case for re-encoding records for Java clients. In fact, we should always re-encode here because the prior call to `get_records()` gives us automatically decoded records, so we must re-encode them before returning them to the client.

### CODE REVIEW

**Comment by SwatsonCodes** [comment]
tests are failing but AFAICT this is unrelated to my changes

File: localstack/services/kinesis/kinesis_listener.py [hunk 1]
```diff
@@ -334,9 +334,6 @@ def subscribe_to_shard(data, headers):
     iterator = kinesis.get_shard_iterator(
         StreamName=stream_name, ShardId=data["ShardId"], ShardIteratorType=iter_type, **kwargs
     )["ShardIterator"]
-    data_needs_encoding = False
-    if "java" in headers.get("User-Agent", "").split(" ")[0]:
```
Comment (whummer) [inline localstack/services/kinesis/kinesis_listener.py:338]: Thanks for digging into this @SwatsonCodes ! 👍 

Looks like we introduced this switch in here for compatibility with the Kinesis Java SDK in [this PR](https://github.com/localstack/localstack/pull/3758). The original tests were introduced in our java-utils repo here: https://github.com/localstack/localstack-java-utils/pull/47/files - should we give it a short try to see if these Java tests are still passing with the changes?

Comment (SwatsonCodes) [reply localstack/services/kinesis/kinesis_listener.py:338]: good thinking, I'll give it a shot.

Comment (whummer) [reply localstack/services/kinesis/kinesis_listener.py:338]: As discussed in person, been able to verify that these changes work against the `KinesisSchedulerTest` Java test. 👍 
```
mvn -Dtest=cloud.localstack.awssdkv2.KinesisSchedulerTest test
...
[INFO] BUILD SUCCESS
```
Not sure if our Java tests are currently actually covering the `SubscribeToShard` API - perhaps that is something we could add at some point..
